File: week2/task3/python/URLCache.py
from node import Node

class URLCache:
    def __init__(self, capacity):
        self.capacity = capacity
        self.map = {}
        self.head = None
        self.tail = None

    def get(self, url):
        if url in self.map:
            node = self.map[url]
            self.remove(node)
            self.setHead(node)
            return node.webPage
        return None

    def set(self, url, webPage):
        if url in self.map:
            old = self.map[url]
            old.webPage = webPage
            self.remove(old)
            self.setHead(old)
        else:
            created = Node(url, webPage)
            if len(self.map) >= self.capacity:
                del self.map[self.tail.url]
                self.remove(self.tail)
                self.setHead(created)
            else:
                self.setHead(created)
            self.map[url] = created

    def remove(self, node):
        if node.prev is not None:
            node.prev.next = node.next
        else:
            self.head = node.next

        if node.next is not None:
            node.next.prev = node.prev
        else:
            self.tail = node.prev

    def setHead(self, node):
        node.next = self.head
        node.prev = None

        if self.head is not None:
            self.head.prev = node

        self.head = node

        if self.tail is None:
            self.tail = self.head

    def printCache(self):
        currentNode = self.head
        while currentNode is not None:
            print(f"[{currentNode.url}, {currentNode.webPage}]", end="")
            currentNode = currentNode.next
            if currentNode is not None:
                print(" -> ", end="")
        print()

File: week2/task3/java/URLCache.java
package java;
import java.util.HashMap;

// LRU(Least Recently Used)
public class URLCache {
    private final int capacity;
    private HashMap<String, Node> map;
    private Node head;
    private Node tail;

    public URLCache(int capacity) {
        this.capacity = capacity;
        map = new HashMap<>();
    }

    public String get(String url) {
        if (map.containsKey(url)) {
            Node node = map.get(url);
            remove(node);
            setHead(node);
            return node.webPage;
        }
        return null;
    }

    public void set(String url, String webPage) {
        if (map.containsKey(url)) {
            Node old = map.get(url);
            old.webPage = webPage;
            remove(old);
            setHead(old);
        } else {
            Node created = new Node(url, webPage);
            if (map.size() >= capacity) {
                map.remove(tail.url);
                remove(tail);
                setHead(created);
            } else {
                setHead(created);
            }
            map.put(url, created);
        }
    }

    private void remove(Node node) {
        if (node.prev != null) {
            node.prev.next = node.next;
        } else {
            head = node.next;
        }

        if (node.next != null) {
            node.next.prev = node.prev;
        } else {
            tail = node.prev;
        }
    }

    private void setHead(Node node) {
        node.next = head;
        node.prev = null;

        if (head != null) {
            head.prev = node;
        }

        head = node;

        if (tail == null) {
            tail = head;
        }
    }

    public void printCache() {
        Node currentNode = head;
        while(currentNode != null) {
            System.out.print("[" + currentNode.url + ", " + currentNode.webPage + "]");
            currentNode = currentNode.next;
            if (currentNode != null) {
                System.out.print(" -> ");
            }
        }
        System.out.println();
    }
}
